src/server/generic_family.cc

```diff
@@ -1509,6 +1511,61 @@ OpStatus GenericFamily::OpMove(const OpArgs& op_args, string_view key, DbIndex t
   return OpStatus::OK;
 }
 
+void GenericFamily::RandomKey(CmdArgList args, ConnectionContext* cntx) {
+  const static size_t kMaxAttemps = 3;
+  std::vector<StringVec> keys(shard_set->size());
+  ScanOpts scan_opts;
+  scan_opts.limit = 3;  // number of entries per shard
+  DbContext db_cntx{.db_index = cntx->conn_state.db_index};
+
+  shard_set->RunBriefInParallel(
+      [&](EngineShard* shard) {
+        auto [prime_table, expire_table] = shard->db_slice().GetTables(db_cntx.db_index);
+        StringVec* candidates = &keys[shard->shard_id()];
+        size_t attempts = 0;
+        uint64_t cursor = 0;
+        do {
+          cursor = prime_table->GetRandomCursor().value();
+          OpScan({shard, 0u, db_cntx}, scan_opts, &cursor, candidates);
+        } while (candidates->empty() && ++attempts < kMaxAttemps);
+
+        if (candidates->empty()) {  // try to scan again from the start of the shard
+          OpScan({shard, 0u, db_cntx}, scan_opts, &cursor, candidates);
+        }
+      },
+      [&](ShardId) { return true; });
+
+  absl::BitGen bitgen;
+
+  auto pick_candidate = [&](StringVec* candidates) -> std::optional<string> {
+    if (candidates->empty()) {
+      return std::nullopt;
+    }
+
+    size_t idx = 0;
+    if (auto cand_size = candidates->size() > 1) {
+      idx = absl::Uniform<size_t>(bitgen, 0, cand_size);
+    }
+
+    return std::optional<string>{candidates->at(idx)};
+  };
+
+  auto rnd_shard_idx = absl::Uniform<size_t>(bitgen, 0, keys.size());
```

Comment:
Another proposal:
How about counting how many elements exist in all vectors (just `count += it->size()`), and then selecting a random number in that range, and then finding the relevant item?
It's somewhat simpler, and also easier to reason about I think

Reply:
We might have a different number of items for each vector, so finding the relevant item after taking a random number in that range means I have to loop the vectors again.

Reply:
Indeed, that's what I propose.
Looping over the vector is cheap, it's limited to the number of physical CPUs in the system, so usually <8, and very rarely is it >32.
But your approach is also ok, it was just a proposal.